File: PROYECTO FINAL - Programación III/cpp_system/Benchmark.cpp
#include <iostream>
#include <fstream>
#include <chrono>

#include "HashTable.h"
#include "AVL.h"

using namespace std;
using namespace chrono;

int main() {

    HashTable hash;

    AVL arbol;

    ofstream archivo(
        "cpp_results.csv"
    );

    archivo
    << "language,operation,"
    << "structure,records,time_ms\n";

    // =========================
    // HASH INSERT
    // =========================

    auto inicioHash =
    high_resolution_clock::now();

    for (int i = 0; i < 10000; i++) {

        Student estudiante;

        estudiante.student_id = i;

        estudiante.full_name =
        "Estudiante";

        estudiante.career =
        "Ingenieria";

        estudiante.semester = 1;

        estudiante.gpa = 85;

        estudiante.skill_score = i;

        hash.insertar(estudiante);
    }

    auto finHash =
    high_resolution_clock::now();

    duration<double, milli>
    tiempoHash =
    finHash - inicioHash;

    archivo
    << "C++,insert,HashTable,"
    << "10000,"
    << tiempoHash.count()
    << endl;

    // =========================
    // HASH SEARCH
    // =========================

    auto inicioBusquedaHash =
    high_resolution_clock::now();

    for (int i = 0; i < 10000; i++) {

        hash.buscar(i);
    }

    auto finBusquedaHash =
    high_resolution_clock::now();

    duration<double, milli>
    tiempoBusquedaHash =
    finBusquedaHash
    - inicioBusquedaHash;

    archivo
    << "C++,search,HashTable,"
    << "10000,"
    << tiempoBusquedaHash.count()
    << endl;

    // =========================
    // AVL INSERT
    // =========================

    auto inicioAVL =
    high_resolution_clock::now();

    for (int i = 0; i < 10000; i++) {

        Student estudiante;

        estudiante.student_id = i;

        estudiante.full_name =
        "Estudiante";

        estudiante.career =
        "Ingenieria";

        estudiante.semester = 1;

        estudiante.gpa = 85;

        estudiante.skill_score = i;

        arbol.insertarEstudiante(
            estudiante
        );
    }

    auto finAVL =
    high_resolution_clock::now();

    duration<double, milli>
    tiempoAVL =
    finAVL - inicioAVL;

    archivo
    << "C++,insert,AVL,"
    << "10000,"
    << tiempoAVL.count()
    << endl;

    archivo.close();

    cout
    << "Benchmark generado "
    << "correctamente\n";

    return 0;
}

File: PROYECTO FINAL - Programación III/Benchmark_java/Benchmark.java
import java.util.HashMap;
import java.util.TreeMap;
import java.io.FileWriter;
import java.io.IOException;

public class Benchmark {

    public static void main(String[] args) {

        try {

            HashMap<Integer, String> hashMap =
                    new HashMap<>();

            TreeMap<Integer, String> treeMap =
                    new TreeMap<>();

            FileWriter archivo =
                    new FileWriter("java_results.csv");

            archivo.write(
                    "language,operation,structure,records,time_ms\n"
            );

            // ==========================
            // HASHMAP INSERT
            // ==========================

            long inicioHash =
                    System.nanoTime();

            for (int i = 0; i < 10000; i++) {

                hashMap.put(
                        i,
                        "Estudiante"
                );
            }

            long finHash =
                    System.nanoTime();

            double tiempoHash =

                    (finHash - inicioHash)
                    / 1000000.0;

            archivo.write(
                    "Java,insert,HashMap,10000,"
                    + tiempoHash + "\n"
            );

            // ==========================
            // HASHMAP SEARCH
            // ==========================

            long inicioBusquedaHash =
                    System.nanoTime();

            for (int i = 0; i < 10000; i++) {

                hashMap.get(i);
            }

            long finBusquedaHash =
                    System.nanoTime();

            double tiempoBusquedaHash =

                    (finBusquedaHash
                    - inicioBusquedaHash)
                    / 1000000.0;

            archivo.write(
                    "Java,search,HashMap,10000,"
                    + tiempoBusquedaHash + "\n"
            );

            // ==========================
            // TREEMAP INSERT
            // ==========================

            long inicioTree =
                    System.nanoTime();

            for (int i = 0; i < 10000; i++) {

                treeMap.put(
                        i,
                        "Estudiante"
                );
            }

            long finTree =
                    System.nanoTime();

            double tiempoTree =

                    (finTree - inicioTree)
                    / 1000000.0;

            archivo.write(
                    "Java,insert,TreeMap,10000,"
                    + tiempoTree + "\n"
            );

            archivo.close();

            System.out.println(
                    "Benchmark Java generado correctamente."
            );

        } catch (IOException e) {

            System.out.println(
                    "Error: "
                    + e.getMessage()
            );
        }
    }
}